### ISSUE TITLE
Mypy should report errors when bpo-33346 would occur

### ISSUE THREAD

**Issue by jhance** [python/mypy#14256]
https://bugs.python.org/issue33346 details some cases where you would get

```
SyntaxError: asynchronous comprehension outside of an asynchronous function
```

in some versions of Python, but mypy doesn't complain about these cases. Arguably this isn't necessarily specific to type checking, but it would be nice to report an error if Python is going to give a syntax error (even though it parses!)

I'm not quite sure what version of Python the fix (https://github.com/python/cpython/commit/054e9c84ac7c394941bba3ea1829d14dce1243fc) is in, but it would be nice to give an error message like "await cannot be used in dict comprehensions" in older versions of Python - especially because the error Python gives at runtime is not so good.

**Comment by JelleZijlstra** [comment]
The fix is in 3.11 and up (you can see the list of tags that the commit is included in on the GH commit page).

**Comment by erictraut** [comment]
This seems like something that would be more appropriate for a linter, since it's not related to type checking. It's also pretty esoteric.
